fastlane_bot/events/listener.py

```diff
@@ -0,0 +1,72 @@
+import asyncio
+import time
+from typing import AsyncGenerator, List
+
+from web3 import AsyncWeb3
+
+from .utils import complex_handler
+from .managers.base import BaseManager
+from .interfaces.event import Event
+
+
+class EventListener:
+    """
+    The EventListener is the entry point to create & manage websocket event subscriptions.
+    """
+    NEW_EVENT_TIMEOUT = 0.01
+
+    def __init__(
+        self,
+        manager: BaseManager,
+        base_exchanges: List[str],
+        w3: AsyncWeb3,
+    ):
+        """ Initializes the EventManager.
+        Args:
+            base_exchanges: The list of base exchanges for which to gather events.
+            carbon_controller_addresses: The list of Carbon Controller addresses for which to gather events.
+            w3: The connected AsyncWeb3 object.
+        """
+        self._manager = manager
+        self._w3 = w3
+        self._subscriptions = []  # Initialize the subscription list
+        self._subscription_by_id = {}
+        self._event_buffer = []
+        self._last_event_ts = 0
+
+        for exchange_name, exchange in self._manager.exchanges.items():
+            if exchange_name in base_exchanges:
+                self._subscriptions += exchange.get_subscriptions(self._manager.event_contracts[exchange_name])
+
```

Comment:
isn't this just essentially
```
self._subscriptions = [
exchange.get_subscriptions(self._manager.event_contracts[exchange_name]
for exchange_name in set(self._manager.exchanges).intersect(base_exchanges)
]
```

Reply:
You would need to flatten it as well

Reply:
It's equivalent (without the need for an additional flattening) to:
```
self._subscriptions = [
    exchange.get_subscriptions(self._manager.event_contracts[exchange_name])
    for exchange_name, exchange in self._manager.exchanges.items()
    if exchange_name in base_exchanges
]
```